src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -17,52 +18,157 @@
+#define GET_COUNT(_args, _size)                                                \
+    ({                                                                         \
+        size_t _count = 0;                                                     \
+        switch ((_args)->coll_type) {                                          \
+        case UCC_COLL_TYPE_ALLREDUCE:                                          \
+            _count = (_args)->dst.info.count;                                  \
+            break;                                                             \
+        case UCC_COLL_TYPE_REDUCE_SCATTER:                                     \
+            _count = UCC_IS_INPLACE(*(_args))                                  \
+                         ? (_args)->dst.info.count                             \
+                         : (_args)->dst.info.count * (_size);                  \
+            break;                                                             \
+        case UCC_COLL_TYPE_REDUCE_SCATTERV:                                    \
+            _count = ucc_coll_args_get_total_count(                            \
+                (_args), (_args)->dst.info_v.counts, (_size));                 \
+        default:                                                               \
+            break;                                                             \
+        }                                                                      \
+        _count;                                                                \
+    })
+
+#define GET_DATA_BUF(_args)                                                    \
+    ({                                                                         \
+        void *_buf = NULL;                                                     \
+        switch ((_args)->coll_type) {                                          \
+        case UCC_COLL_TYPE_ALLREDUCE:                                          \
+        case UCC_COLL_TYPE_REDUCE_SCATTER:                                     \
+            _buf = UCC_IS_INPLACE(*(_args)) ? (_args)->dst.info.buffer         \
+                                            : (_args)->src.info.buffer;        \
+            break;                                                             \
+        case UCC_COLL_TYPE_REDUCE_SCATTERV:                                    \
+            _buf = UCC_IS_INPLACE(*(_args)) ? (_args)->dst.info_v.buffer       \
+                                            : (_args)->src.info_v.buffer;      \
+        default:                                                               \
+            break;                                                             \
+        }                                                                      \
+        _buf;                                                                  \
+    })
+
+#define GET_DT(_args)                                                          \
```

Comment:
these functions seem repetitive to what you did in allgather_knomial.c - perhaps make a more general function in tl_ucp_coll.h?